Mathlib/Combinatorics/Enumerative/Bell.lean

```diff
@@ -0,0 +1,184 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir & María-Inés de Frutos—Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María-Inés de Frutos—Fernández
+-/
+
+import Mathlib.Data.Nat.Choose.Multinomial
+import Mathlib.Data.Nat.Choose.Mul
+
+/-! # Bell numbers for multisets
+
+For `n : ℕ`, the `n`th Bell number is the number of partitions of a set of cardinality `n`.
+Here, we define a refinement of these numbers, that count, for any `m : Multiset ℕ`,
+the number of partitions of a set of cardinality `m.sum` whose parts have cardinalities
+given by `m`.
+
+The definition presents it as an integer.
+
+* `Multiset.bell`: number of partitions of a set whose parts have cardinalities a given multiset
+
+* `Nat.uniformBell m n` : short name for `Multiset.bell (replicate m n)`
+
+* `Multiset.bell_mul_eq` shows that
+    `m.bell * (m.map (fun j ↦ j !)).prod *
+      Π j ∈ (m.toFinset.erase 0), (m.count j)! = m.sum !`
+
+* `Nat.uniformBell_mul_eq`  shows that
+    `uniformBell m n * n ! ^ m * m ! = (m * n)!`
+
+* `Nat.uniformBell_succ_left` computes `Nat.uniformBell (m + 1) n` from `Nat.uniformBell m n`
+
+## TODO
+
+Prove that it actually counts the number of partitions as indicated.
+(When `m` contains `0`, the result requires to admit repetitions of the empty set as a part.)
+
+-/
+
+open Multiset Nat
+
+namespace Multiset
+
+/-- Number of partitions of a set of cardinality `m.sum`
+whose parts have cardinalities given by `m` -/
+def bell (m : Multiset ℕ) : ℕ :=
+  Nat.multinomial m.toFinset (fun k ↦ k * m.count k) *
+    ∏ k ∈ m.toFinset.erase 0, ∏ j ∈ .range (m.count k), (j * k + k - 1).choose (k - 1)
+
+@[simp]
+theorem bell_zero : bell 0 = 1 := rfl
+
+private theorem bell_mul_eq_lemma {x : ℕ} (hx : x ≠ 0) (c : ℕ) :
+    x ! ^ c * c ! * ∏ j ∈ Finset.range c, (j * x + x - 1).choose (x - 1) = (x * c)! := by
+  induction c with
+  | zero => simp
+  | succ c hrec =>
+    suffices  x ! ^ (c + 1) * (c + 1) ! = x ! * (c + 1) * (x ! ^ c * c !) by
+      rw [this]
+      rw [← mul_assoc]
+      simp only [Finset.range_succ, Finset.mem_range, lt_self_iff_false, not_false_eq_true,
+        Finset.prod_insert]
+      simp only [mul_assoc]
+      rw [mul_comm ((c * x + x - 1).choose (x-1))]
+      rw [← mul_assoc]
+      rw [mul_comm]
+      simp only [← mul_assoc]
+      rw [hrec]
+      have : (x * c)! * ((c * x + x - 1).choose (x - 1)) * x ! * (c + 1)
+        = ((c + 1) * ((c * x + x - 1).choose (x - 1))) * (x * c)! *  x ! := by
+        ring_nf
+      rw [this]
+      rw [← Nat.choose_mul_add c hx]
+      rw [mul_comm c x]
+      rw [Nat.add_choose_mul_factorial_mul_factorial (x * c) x]
+      rw [mul_add, mul_one]
+    rw [factorial_succ, pow_succ]
+    ring_nf
```

Comment:
This looks like a good candidate for a `calc` proof. Do you mind trying?

Reply:
done